File: code/compiler/btree/tree.go
package main

import "fmt"

var (
	tree = &Tree{"a",
		&Tree{"b",
			&Tree{"d", nil, nil},
			nil},
		&Tree{"c",
			nil,
			&Tree{"e", nil, nil},
		},
	}
)

// Tree : binary tree
type Tree struct {
	value string
	left  *Tree
	right *Tree
}

func (tree *Tree) String() string {
	if tree == nil {
		return ""
	}
	return fmt.Sprintf("%s{%s,%s}",
		tree.value, tree.left, tree.right)
}

// Reverse binary tree
func Reverse(tree *Tree) {
	if tree == nil {
		return
	}

	t := tree.left
	tree.left = tree.right
	tree.right = t

	Reverse(tree.left)
	Reverse(tree.right)
}

func main() {
	fmt.Println("[ORIGINAL]:\t\t", tree)
	Reverse(tree)
	fmt.Println("[FLIPPED]:\t\t", tree)
}

File: code/compiler/btree/tree.js
const println = console.log

function Btree(value, left, right) {
    this.value = value
    this.left = left
    this.right = right
}

function ReverseTree(tree) {
    if (!tree) {
        return
    }
    t = tree.left
    tree.left = tree.right
    tree.right = t

    ReverseTree(tree.left)
	ReverseTree(tree.right)
}

const tree = new Btree(10, new Btree(20, new Btree(30)), new Btree(30, null, new Btree(30)))
ReverseTree(tree)
println(tree)
